### PR TITLE
BATCH-2041: Add support for reading excel files

### PR DESCRIPTION
**Pull request by mdeinum**
This commit adds support for reading Excel files. Both JExcel and Apache POI are supported. Support for reading Excel files has been designed similar to the support for reading flat files.

I have signed and agree to the terms of the SpringSource Individual Contributor License Agreement.

### CODE REVIEW

File: spring-batch-excel/src/main/java/org/springframework/batch/item/excel/poi/PoiSheet.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        if (row == null) {
+            return null;
+        }
+        final List<String> cells = new LinkedList<String>();
+
+        for (int i = 0; i < getNumberOfColumns(); i++) {
+            Cell cell = row.getCell(i);
+            switch (cell.getCellType()) {
+                case Cell.CELL_TYPE_NUMERIC:
+                    cells.add(String.valueOf(cell.getNumericCellValue()));
```
Comment (andrewdmoreno) [inline spring-batch-excel/src/main/java/org/springframework/batch/item/excel/poi/PoiSheet.java]: Been checking this extension out for use in a project and was just wondering if there was a reason the `CELL_TYPE_NUMERIC` case does not check for a Date formatted numeric as shown in the POI quick guide (see http://poi.apache.org/spreadsheet/quick-guide.html#CellContents). Without this the date is returned as the serialized number. Wasn't sure if this was the intention so thought I'd mention it.

From POI guide:

```
case Cell.CELL_TYPE_NUMERIC:
    if (DateUtil.isCellDateFormatted(cell)) {
        System.out.println(cell.getDateCellValue());
    } else {
        System.out.println(cell.getNumericCellValue());
    }
    break;
```

My apologies if this was the incorrect place to note this.

Comment (mdeinum) [inline spring-batch-excel/src/main/java/org/springframework/batch/item/excel/poi/PoiSheet.java]: That was an oversight, thanks for pointing it out.

File: spring-batch-excel/src/main/java/org/springframework/batch/item/excel/poi/PoiItemReader.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @Override
+    protected int getNumberOfSheets() {
+        return this.workbook.getNumberOfSheets();
+    }
+
+    @Override
+    protected void openExcelFile(final Resource resource) throws Exception {
+        this.workbook = WorkbookFactory.create(resource.getInputStream());
```
Comment (mdeinum) [inline spring-batch-excel/src/main/java/org/springframework/batch/item/excel/poi/PoiItemReader.java]: There, sadly, is no `close()` method on the created workbook. However according to what I've read closing the `InputStream` should fix that. However the earlier solution was faulty, as you commented earlier. For now I keep the used `InputStream` and close it in the `doClose` method.

**Comment by dusiema** [comment]
Will this be merged any time soon? What is the current status of this?

**Comment by mminella** [comment]
Unless you have any objections, I'm going to merge this with one small update which is to update the `@Depricated` notes to point to the POI implementation (so users know what to use instead).

**Comment by dusiema** [comment]
No, I don't have any. I am using this already as a clone from https://github.com/mdeinum/spring-batch-extensions. While using it I found one or two things that could be improved (like being able to specify just one sheet in an excel file). But I think this could also be done, when it is integrated into this repo.

**Comment by mdeinum** [comment]
@mminella Would this be worth a blog post explaining things a little? 

@dusiema Maybe you could create an issue here at github and provide a pull request if you already have the code (after it has been merged ofcourse).

**Comment by dusiema** [comment]
@mdeinum: I don't have the code yet (I just deleted all the other sheets from the excel file). I will create  an enhancement issue and mabye I can also provide a pull request later on. However before creating the issue in this repo I guess the code should be here first.

**Comment by dusiema** [comment]
@mminella: Any update on the integration? I have two more enhancement suggestions and I could provide pull-requests for them. I just thought I could provide them right here instead of the repo from @mdeinum.

**Comment by mdeinum** [comment]
@dusiema If you have the enhancements you can create 1 or 2 pull requests to my repo. I'll merge them and then they should show up here also.

**Comment by dusiema** [comment]
@mdeinum I already thought about that. Need to get back to my computer at work. I'll try to do it tomorrow.

**Comment by mminella** [comment]
Hey guys, I'm at a conference through tomorrow.  I plan on merging this Friday.

**Comment by dusiema** [comment]
@mminella Cool. 
I'll try to do the pull request anyways so @mdeinum can have a look first.

**Comment by mminella** [comment]
Merged.  Thanks for all the hard work!

**Comment by mdeinum** [comment]
Great... Thanks for merging it. I'll try to write a blog post in the next couple of days.

**Pull request closed without merging**